fabtests/unit/getinfo_test.c

```diff
@@ -964,6 +964,259 @@ getinfo_test(caps, 4, "Test for capability bit regression",
+	 * This capabilites should be set in the return fi_info::caps. In
+	 * addition, since user hint domain_attr::caps, tx_attr::caps, and
+	 * rx_attr::caps are zero, these returns caps should be set based on
+	 * fi_info::caps
+	 */
+	hints->caps = FI_TAGGED;
+	hints->optional_caps = FI_MSG | FI_MULTI_RECV | FI_LOCAL_COMM |
+		FI_REMOTE_COMM;
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with optional_caps failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, required_caps,
+					   required_caps & DOMAIN_CAPS,
+					   required_caps & (PRIMARY_TX_CAPS | OPT_TX_CAPS),
+					   required_caps & (PRIMARY_RX_CAPS | OPT_RX_CAPS));
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	/* Verify that using domain_attr::optional_caps, tx_attr::optional_caps,
+	 * and rx_attr::optional_caps works as expected.
+	 */
+	hints->caps = required_caps;
+	hints->optional_caps = 0;
+	hints->domain_attr->caps = FI_REMOTE_COMM;
+	hints->tx_attr->caps = FI_TAGGED;
+	hints->rx_attr->caps = FI_TAGGED;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, required_caps, FI_REMOTE_COMM,
+					   FI_TAGGED, FI_TAGGED);
+	if (ret != FI_SUCCESS)
+		return ret;
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	hints->domain_attr->optional_caps = FI_LOCAL_COMM;
+	hints->tx_attr->optional_caps = FI_MSG | FI_MULTI_RECV;
+	hints->rx_attr->optional_caps = FI_MSG | FI_MULTI_RECV;
+
+	ret = fi_getinfo(FI_VERSION(1, 20), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("fi_getinfo with optional_caps failed to return\n");
+		goto out;
+	}
+
+	ret = test_valid_optional_cap_info(*info, required_caps,
+					   required_caps & DOMAIN_CAPS,
+					   required_caps & (PRIMARY_TX_CAPS | OPT_TX_CAPS),
+					   required_caps & (PRIMARY_RX_CAPS | OPT_RX_CAPS));
+	if (ret != FI_SUCCESS)
+		return ret;
+
+out:
+	return ret;
+}
+
+static int test_ignore_optional_cap(char *node, char *service, uint64_t flags,
+				    struct fi_info *hints,
+				    struct fi_info **info)
+{
+	uint64_t required_caps = FI_MSG | FI_TAGGED | FI_MULTI_RECV |
+		FI_LOCAL_COMM | FI_REMOTE_COMM;
+	int ret;
+
+	/* Verify that the provider meets the required capabilities of this
+	 * test.
+	 */
+	hints->caps = required_caps;
+	ret = fi_getinfo(FI_VERSION(1, 19), node, service, flags, hints, info);
+	if (ret != FI_SUCCESS) {
+		FT_DEBUG("No available providers supporting FI_MSG | FI_RMA\n");
+		goto out;
+	}
+
+	hints->domain_attr->name = strdup((*info)->domain_attr->name);
+
+	fi_freeinfo(*info);
+	*info = NULL;
+
+	if (!hints->domain_attr->name) {
+		ret = -FI_ENOMEM;
+		goto out;
+	}
+
+	/* Verify that all optional caps are ignore since requested API version
+	 * is less than 1.20.
```

Comment:
This isn't quite true.  The provider may only ignore primary caps.  It can still enable secondary caps, even if it never checks the optional_caps field.

This brings up a point that the optional_caps field really only applies to primary caps.

Reply:
I'll update this test to pass in only primary caps.

From API point-of-view, I think we still want to allow for secondary caps to be passed in via optional_caps. For example, if a provider does not support FI_SOURCE_ERR and the app supports either with or without FI_SOURCE_ERR, having the app pass this in via optional_caps would avoid multiple calls to fi_getinfo().

Reply:
It's fine to allow the app to pass in secondary caps, but it doesn't make a difference from the app's viewpoint.  Secondary caps an always be returned even if not requested.  They're kind of always optional anyway.

Reply:
The difference is that if app asks for secondary caps via the regular caps and the provider doesn't support it fi_getinfo() would fail.  A provider may choose not to report a secondary caps unless the application asks for it due to the overhead associated with supporting it. By passing via optional caps both scenarios are handled.

Reply:
That difference should be made clear to the app then.  If requested as optional, it will be returned if supported, even if there is a performance penalty for supporting it.

Reply:
Right, if requested as optional, it will be returned if supported and still succeed if not supported. However, if requested as normal caps, it will fail if not supported.

Reply:
I was referring to the cases of not being requested versus being requested as optional.

If the app is concerned about performance setting the secondary capability bit as an optional_cap is a worse choice than leaving it unset.  That result is not obvious and likely counterintuitive to a developer.  Since the bit is marked as optional, the app can obviously work around not having that capability.

Maybe we need a crisper definition of how a provider interprets an optional_cap.  Do we say that the provider MUST support the cap if it can, or that it MAY support the cap if it can?  And, if the later, what conditions do we allow the provider to say no?

It seems weird that a secondary capability bit suddenly becomes a mandatory requirement of the provider if set as an optional capability (assuming the provider supports the feature).

Reply:
I think it is still up to the provider to decide whether it should report the support of an optional caps or not. It is also allowed that the provider may behave differently (or identically) for the case of a secondary cap not being requested vs requested as optional.

Reply:
One of the main use-cases for optional_caps would be apps can discover if a provider supports a capability without having to go through multiple iterations of passing in different or NULL hints to fi_getinfo(). For example, since FI_SOURCE is typically not reported by default from providers (even if supported), this could be a good way for users to discover if providers support FI_SOURCE. It would still be up to users if they want to use FI_SOURCE when allocating endpoints. Assuming since they have asked for it, they understand the potential performance implications.

>I think it is still up to the provider to decide whether it should report the support of an optional caps or not.

Allowing providers to selectively support optional_caps may result in users reverting to multiple iterations of hints and fi_getinfo() to discover features.

Reply:
I like Ian's suggestion that apps can then set the caps that they want based on what's returned as an optional_caps.  That would still need to be called out, since optional_caps would still be ignored on object allocation.

Reply:
Looks good to me.

Reply:
This works for me.